proto/message.go

```diff
@@ -201,13 +201,19 @@ func Parse(m RawMessage) (FieldNumber, WireType, RawValue, RawMessage, error) {
 		if err != nil {
 			return f, t, nil, m, fmt.Errorf("decoding varint field %d: %w", f, err)
 		}
+		if len(m) < n {
+			return f, t, nil, m, fmt.Errorf("decoding varint field %d: %w", f, io.ErrUnexpectedEOF)
+		}
 		return f, t, RawValue(m[:n]), m[n:], nil
 
 	case Varlen:
 		l, n, err := decodeVarint(m) // length
 		if err != nil {
 			return f, t, nil, m, fmt.Errorf("decoding varlen field %d: %w", f, err)
 		}
+		if len(m) < (n + int(l)) {
```

Comment:
I think `uint64(len(m) - n) < l` would be safer here. You can use `math.MaxUint64` (encoded as a varint: `append(bytes.Repeat([]byte{0xFF}, 9), 0x1)`) as a test vector here. Without this check, `int(l)` will be `-1` and it'll panic below.

Reply:
Oh computers...